registration-processor/pre-processor/registration-processor-cmd-validator-stage/src/main/java/io/mosip/registration/processor/stages/cmdvalidator/DeviceValidator.java

```diff
@@ -62,81 +77,156 @@
-				LocalDateTime localdatetime = LocalDateTime
-						.parse(DateUtils.getUTCCurrentDateTimeString(env.getProperty(DATETIME_PATTERN)), format);
-				request.setRequesttime(localdatetime);
-
-				DeviceValidateHistoryResponse deviceValidateResponse;
-
-				ResponseWrapper<?> responseWrapper = (ResponseWrapper<?>) registrationProcessorRestService
-						.postApi(ApiName.DEVICEVALIDATEHISTORY, "", "", request, ResponseWrapper.class);
-				deviceValidateResponse = mapper.readValue(mapper.writeValueAsString(responseWrapper.getResponse()),
-						DeviceValidateHistoryResponse.class);
-				if (responseWrapper.getErrors() == null || responseWrapper.getErrors().isEmpty()) {
-					if (!deviceValidateResponse.getStatus().equalsIgnoreCase(VALID)) {
-						throw new ValidationFailedException(StatusUtil.DEVICE_VALIDATION_FAILED.getMessage(),
-								StatusUtil.DEVICE_VALIDATION_FAILED.getCode());
-					}
-				} else {
-					List<ErrorDTO> error = responseWrapper.getErrors();
-					regProcLogger.error("validate call ended for registrationId {} with error data : {}",
-							registrationId, error.get(0).getMessage());
-					throw new BaseCheckedException(
-							error.get(0).getMessage() + " for " + deviceDetails.getDigitalId().getType(),
-							StatusUtil.DEVICE_VALIDATION_FAILED.getCode());
-
+	public void validate(RegOsiDto regOsi,String process, String registrationId)
+			throws JsonProcessingException, IOException, BaseCheckedException, ApisResourceAccessException, JSONException {
+
+		BiometricRecord biometricRecord = packetManagerService.getBiometricsByMappingJsonKey(registrationId,
+				MappingJsonConstants.INDIVIDUAL_BIOMETRICS, process,
+				ProviderStageName.CMD_VALIDATOR);
+		List<BIR> birs=biometricRecord.getSegments();
+		List<JSONObject> payloads=new ArrayList<>();
+		for(BIR bir: birs) {
+			if(bir.getOthers()!=null) {
+			for(Entry entry: bir.getOthers()) {
+				if(entry.getKey().equals("PAYLOAD")) {
+					payloads.add(new JSONObject(entry.getValue()));				
 				}
-				regProcLogger.debug("validate call ended for registrationId {} with response data : {}", registrationId,
-						JsonUtil.objectMapperObjectToJson(deviceValidateResponse));
 			}
+			}
+		}
+		if(payloads==null || payloads.isEmpty()) {
+			throw new BaseCheckedException(
+					StatusUtil.DEVICE_VALIDATION_FAILED.getCode(),StatusUtil.DEVICE_VALIDATION_FAILED.getMessage()+"-->Others info is not prsent in packet");
+		}else {
+			for(JSONObject payload :payloads) {
+				validateDigitalId(payload) ;
+				
+				String digitalIdString=new String(CryptoUtil.decodeBase64(payload.getString("digitalId").split("\\.")[1]));
+				NewDigitalId newDigitalId=mapper.readValue(digitalIdString, NewDigitalId.class);
+				
+				validateDigitalIdTimestamp (payload,regOsi.getPacketCreationDate(),newDigitalId.getDateTime()) ;
+				
+				validateDeviceForHotlist(payload.getString("deviceCode"),newDigitalId.getDateTime());
```

Comment:
for both validateDigitalId and validateDeviceForHotlist can we take the unique set of digitalId and deviceCode and do this once for unique to avoid unwanted repeat validation?

The most repeat case is fingers, there will be 10 payload all having same digitalId and deviceCode and performing 10 times is a waste of resource.

Reply:
The digittal id might be same , but signatures might differ. For hotlist validation we are also checking if the device is blocked till particular time period or not, so the validation happens for both device code and timestamp .

Reply:
The signature of the payload will be different, but the signature within the digital id will not change.
Most of the cases timestamp will also be same, so you can take a unique combination of deviceCode, timestamp and do hotlist validation for the same. This will also help in faster processing of the packet